### PR TITLE
feat: add mercurial segment

### PR DESCRIPTION
**Pull request by davidcourtney**
### Prerequisites

- [X] I have read and understood the [contributing guide][CONTRIBUTING.md]
- [X] The commit message follows the [conventional commits][cc] guidelines
- [X] Tests for the changes have been added (for bug fixes/features)
- [X] Docs have been added/updated (for bug fixes/features)

### Description

Added a segment for the [Mercurial][mercurial] source control system

[mercurial]: https://www.mercurial-scm.org
[CONTRIBUTING.md]: https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/CONTRIBUTING.md
[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

### CODE REVIEW

File: src/segments/mercurial.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+func (hg *Mercurial) Enabled() bool {
+	if !hg.shouldDisplay() {
+		return false
+	}
+
+	displayStatus := hg.props.GetBool(FetchStatus, false)
+	if displayStatus {
+		hg.setMercurialStatus()
+	} else {
+		hg.Working = &MercurialStatus{}
```
Comment (JanDeDobbeleer) [inline src/segments/mercurial.go]: you could initialise this above and drop the initialisation at line 97. In that case you don't need and `if else`.

File: src/segments/mercurial.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	hg.Dir = strings.TrimSuffix(dir, "/.hg")
+}
+
+func (hg *Mercurial) setMercurialStatus() {
+	hg.Working = &MercurialStatus{}
+	hg.Branch = hg.command
+
+	idString := hg.getHgCommandOutput("log", "-r", ".", "--template", hgLogTemplate)
+	if idString == "" {
```
Comment (JanDeDobbeleer) [inline src/segments/mercurial.go]:
```suggestion
	if len(idString) == 0 {
```

for consistency with everything else

File: src/segments/mercurial.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	idSplit := strings.Split(idString, "|")
+	if len(idSplit) != 5 {
+		return
+	}
+
+	hg.LocalCommitNumber = idSplit[0]
+	hg.ChangeSetID = idSplit[1]
+	hg.ChangeSetIDShort = idSplit[1][:12]
```
Comment (JanDeDobbeleer) [inline src/segments/mercurial.go]: this length isn't validated, should we do that?

File: src/segments/mercurial.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		}
+	}
+
+	if hg.IsTip {
+		hg.Tags = RemoveAtIndex(hg.Tags, tipIndex)
+	}
+
+	statusString := hg.getHgCommandOutput("status")
+
+	if statusString == "" {
```
Comment (JanDeDobbeleer) [inline src/segments/mercurial.go]:
```suggestion
	if len(statusString) == 0 {
```

File: src/segments/mercurial.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+	statusLines := strings.Split(statusString, "\n")
+
+	for _, status := range statusLines {
+		hg.Working.add(status[:1])
+	}
+}
+
+func doSplit(s string) []string {
+	if s == "" {
```
Comment (JanDeDobbeleer) [inline src/segments/mercurial.go]:
```suggestion
	if len(s) == 0 {
```

**Review by JanDeDobbeleer — CHANGES_REQUESTED** [review]

**Comment by davidcourtney** [comment]
I made all the suggested changes and rebased the branch commit from the latest in main.

**Review by JanDeDobbeleer — APPROVED** [review]

**Comment by JanDeDobbeleer** [comment]
@all-contributors please add @davidcourtney for code,design,doc

**Pull request merged**